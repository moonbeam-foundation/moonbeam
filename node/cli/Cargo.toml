[package]
name = "moonbeam-cli"
authors = { workspace = true }
edition = "2021"
version = "0.31.0"

[dependencies]
clap = { workspace = true, features = [ "derive" ] }
log = { workspace = true }
parity-scale-codec = { workspace = true }

# Moonbeam
moonbeam-cli-opt = { workspace = true }
moonbeam-service = { workspace = true }

# Substrate
frame-benchmarking-cli = { workspace = true }
sc-cli = { workspace = true }
sc-finality-grandpa = { workspace = true }
sc-service = { workspace = true }
sc-sysinfo = { workspace = true }
sc-telemetry = { workspace = true }
sc-tracing = { workspace = true }
sp-core = { workspace = true, features = [ "std" ] }
sp-io = { workspace = true }
sp-runtime = { workspace = true, features = [ "std" ] }
sp-wasm-interface = { workspace = true }
substrate-prometheus-endpoint = { workspace = true }
try-runtime-cli = { workspace = true, optional = true }

# Cumulus / Nimbus
cumulus-client-cli = { workspace = true }
cumulus-client-service = { workspace = true }
cumulus-primitives-core = { workspace = true, features = [ "std" ] }
nimbus-primitives = { workspace = true, features = [ "std" ] }

# Polkadot
polkadot-cli = { workspace = true }
polkadot-parachain = { workspace = true, features = [ "std" ] }
polkadot-primitives = { workspace = true }
polkadot-service = { workspace = true }

[build-dependencies]
substrate-build-script-utils = { workspace = true }

[features]
default = [ "moonbase-native", "moonbeam-native", "moonriver-native", "westend-native" ]

westend-native = [ "polkadot-service/westend-native" ]

moonbase-native = [ "moonbeam-service/moonbase-native", "westend-native" ]
moonbeam-native = [ "moonbeam-service/moonbeam-native" ]
moonriver-native = [ "moonbeam-service/moonriver-native" ]

runtime-benchmarks = [ "moonbeam-service/runtime-benchmarks", "polkadot-service/runtime-benchmarks" ]
try-runtime = [
	"moonbeam-service/try-runtime",
	"try-runtime-cli/try-runtime",
]

moonbase-runtime-benchmarks = [ "moonbeam-service/moonbase-runtime-benchmarks" ]
