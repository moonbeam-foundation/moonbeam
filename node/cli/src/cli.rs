// Copyright 2019-2022 PureStake Inc.
// This file is part of Moonbeam.

// Moonbeam is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Moonbeam is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.

//! Moonbeam CLI Library. Built with clap
//!
//! This module defines the Moonbeam node's Command Line Interface (CLI)
//! It is built using clap and inherits behavior from Substrate's sc_cli crate.

use clap::Parser;
use moonbeam_cli_opt::{account_key::GenerateAccountKey, EthApi, Sealing};
use moonbeam_service::chain_spec;
use sc_cli::{Error as CliError, SubstrateCli};
use std::path::PathBuf;

/// Sub-commands supported by the collator.
#[derive(Debug, clap::Subcommand)]
pub enum Subcommand {
	/// Export the genesis state of the parachain.
	#[clap(name = "export-genesis-state")]
	ExportGenesisState(ExportGenesisStateCommand),

	/// Export the genesis wasm of the parachain.
	#[clap(name = "export-genesis-wasm")]
	ExportGenesisWasm(ExportGenesisWasmCommand),

	/// Build a chain specification.
	BuildSpec(BuildSpecCommand),

	/// Validate blocks.
	CheckBlock(sc_cli::CheckBlockCmd),

	/// Export blocks.
	ExportBlocks(sc_cli::ExportBlocksCmd),

	/// Export the state of a given block into a chain spec.
	ExportState(sc_cli::ExportStateCmd),

	/// Import blocks.
	ImportBlocks(sc_cli::ImportBlocksCmd),

	/// Remove the whole chain.
	PurgeChain(cumulus_client_cli::PurgeChainCmd),

	/// Revert the chain to a previous state.
	Revert(sc_cli::RevertCmd),

	/// Sub-commands concerned with benchmarking.
	/// The pallet benchmarking moved to the `pallet` sub-command.
	#[clap(subcommand)]
	Benchmark(frame_benchmarking_cli::BenchmarkCmd),

	/// Try some command against runtime state.
	#[cfg(feature = "try-runtime")]
	TryRuntime(try_runtime_cli::TryRuntimeCmd),

	/// Try some command against runtime state. Note: `try-runtime` feature must be enabled.
	#[cfg(not(feature = "try-runtime"))]
	TryRuntime,

	/// Key management cli utilities
	#[clap(subcommand)]
	Key(KeyCmd),
}

#[derive(Debug, Parser)]
pub struct BuildSpecCommand {
	#[clap(flatten)]
	pub base: sc_cli::BuildSpecCmd,

	/// Number of accounts to be funded in the genesis
	/// Warning: This flag implies a development spec and overrides any explicitly supplied spec
	#[clap(long, conflicts_with = "chain")]
	pub accounts: Option<u32>,

	/// Mnemonic from which we can derive funded accounts in the genesis
	/// Warning: This flag implies a development spec and overrides any explicitly supplied spec
	#[clap(long, conflicts_with = "chain")]
	pub mnemonic: Option<String>,
}

/// Command for exporting the genesis state of the parachain
#[derive(Debug, Parser)]
pub struct ExportGenesisStateCommand {
	/// Output file name or stdout if unspecified.
	#[clap(value_parser)]
	pub output: Option<PathBuf>,

	/// Id of the parachain this state is for.
	#[clap(long)]
	pub parachain_id: Option<u32>,

	/// Write output in binary. Default is to write in hex.
	#[clap(short, long)]
	pub raw: bool,

	/// The name of the chain for that the genesis state should be exported.
	#[clap(long)]
	pub chain: Option<String>,
}

/// Command for exporting the genesis wasm file.
#[derive(Debug, Parser)]
pub struct ExportGenesisWasmCommand {
	/// Output file name or stdout if unspecified.
	#[clap(value_parser)]
	pub output: Option<PathBuf>,

	/// Write output in binary. Default is to write in hex.
	#[clap(short, long)]
	pub raw: bool,

	/// The name of the chain for that the genesis wasm file should be exported.
	#[clap(long)]
	pub chain: Option<String>,
}

#[derive(Debug, Parser)]
#[group(skip)]
pub struct RunCmd {
	#[clap(flatten)]
	pub base: cumulus_client_cli::RunCmd,

	/// Enable the development service to run without a backing relay chain
	#[clap(long)]
	pub dev_service: bool,

	/// When blocks should be sealed in the dev service.
	///
	/// Options are "instant", "manual", or timer interval in milliseconds
	#[clap(long, default_value = "instant")]
	pub sealing: Sealing,

	/// Public authoring identity to be inserted in the author inherent
	/// This is not currently used, but we may want a way to use it in the dev service.
	// #[clap(long)]
	// pub author_id: Option<NimbusId>,

	/// Enable EVM tracing module on a non-authority node.
	#[clap(
		long,
		conflicts_with = "collator",
		conflicts_with = "validator",
		value_delimiter = ','
	)]
	pub ethapi: Vec<EthApi>,

	/// Number of concurrent tracing tasks. Meant to be shared by both "debug" and "trace" modules.
	#[clap(long, default_value = "10")]
	pub ethapi_max_permits: u32,

	/// Maximum number of trace entries a single request of `trace_filter` is allowed to return.
	/// A request asking for more or an unbounded one going over this limit will both return an
	/// error.
	#[clap(long, default_value = "500")]
	pub ethapi_trace_max_count: u32,

	/// Duration (in seconds) after which the cache of `trace_filter` for a given block will be
	/// discarded.
	#[clap(long, default_value = "300")]
	pub ethapi_trace_cache_duration: u64,

	/// Size in bytes of the LRU cache for block data.
	#[clap(long, default_value = "300000000")]
	pub eth_log_block_cache: usize,

	/// Size in bytes of the LRU cache for transactions statuses data.
	#[clap(long, default_value = "300000000")]
	pub eth_statuses_cache: usize,

	/// Size in bytes of data a raw tracing request is allowed to use.
	/// Bound the size of memory, stack and storage data.
	#[clap(long, default_value = "20000000")]
	pub tracing_raw_max_memory_usage: usize,

	/// Maximum number of logs in a query.
	#[clap(long, default_value = "10000")]
	pub max_past_logs: u32,

	/// Force using Moonbase native runtime.
	#[clap(long = "force-moonbase")]
	pub force_moonbase: bool,

	/// Force using Moonriver native runtime.
	#[clap(long = "force-moonriver")]
	pub force_moonriver: bool,

	/// Id of the parachain this collator collates for.
	#[clap(long)]
	pub parachain_id: Option<u32>,

	/// Maximum fee history cache size.
	#[clap(long, default_value = "2048")]
	pub fee_history_limit: u64,

	/// Disable automatic hardware benchmarks.
	///
	/// By default these benchmarks are automatically ran at startup and measure
	/// the CPU speed, the memory bandwidth and the disk speed.
	///
	/// The results are then printed out in the logs, and also sent as part of
	/// telemetry, if telemetry is enabled.
	#[clap(long)]
	pub no_hardware_benchmarks: bool,
}

impl std::ops::Deref for RunCmd {
	type Target = cumulus_client_cli::RunCmd;

	fn deref(&self) -> &Self::Target {
		&self.base
	}
}

#[derive(Debug, clap::Subcommand)]
pub enum KeyCmd {
	#[clap(flatten)]
	BaseCli(sc_cli::KeySubcommand),
	/// Generate an Ethereum account.
	GenerateAccountKey(GenerateAccountKey),
}

impl KeyCmd {
	/// run the key subcommands
	pub fn run<C: SubstrateCli>(&self, cli: &C) -> Result<(), CliError> {
		match self {
			KeyCmd::BaseCli(cmd) => cmd.run(cli),
			KeyCmd::GenerateAccountKey(cmd) => {
				cmd.run();
				Ok(())
			}
		}
	}
}

#[derive(Debug, Parser)]
#[clap(
	propagate_version = true,
	args_conflicts_with_subcommands = true,
	subcommand_negates_reqs = true
)]
pub struct Cli {
	#[clap(subcommand)]
	pub subcommand: Option<Subcommand>,

	#[clap(flatten)]
	pub run: RunCmd,

	/// Relaychain arguments
	#[clap(raw = true)]
	pub relaychain_args: Vec<String>,
}

#[derive(Debug)]
pub struct RelayChainCli {
	/// The actual relay chain cli object.
	pub base: polkadot_cli::RunCmd,

	/// Optional chain id that should be passed to the relay chain.
	pub chain_id: Option<String>,

	/// The base path that should be used by the relay chain.
	pub base_path: Option<PathBuf>,
}

impl RelayChainCli {
	/// Parse the relay chain CLI parameters using the para chain `Configuration`.
	pub fn new<'a>(
		para_config: &sc_service::Configuration,
		relay_chain_args: impl Iterator<Item = &'a String>,
	) -> Self {
		let extension = chain_spec::Extensions::try_get(&*para_config.chain_spec);
		let chain_id = extension.map(|e| e.relay_chain.clone());
		let base_path = para_config
			.base_path
			.as_ref()
			.map(|x| x.path().join("polkadot"));
		Self {
			base_path,
			chain_id,
			base: polkadot_cli::RunCmd::parse_from(relay_chain_args),
		}
	}
}
