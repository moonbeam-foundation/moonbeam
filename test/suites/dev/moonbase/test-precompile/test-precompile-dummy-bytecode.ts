import "@moonbeam-network/api-augment";
import { describeSuite, expect } from "@moonwall/cli";
import {
  ALITH_ADDRESS,
  DUMMY_REVERT_BYTECODE,
  PRECOMPILE_PARACHAIN_STAKING_ADDRESS,
  createEthersTransaction,
} from "@moonwall/util";
import * as RLP from "rlp";
import { keccak256 } from "viem";

// push1 5 (deployed bytecode length)
// dup1
// push1 11 (offset of deployed bytecode in this initcode)
// push1 0 (offset in target memory)
// codecopy (copy code slice into memory)
// push1 0 (offset in target memory)
// return
// <deployed bytecode>
const INIT_CODE = "0x600580600B6000396000F360006000fd";

describeSuite({
  id: "D2534",
  title: "Precompiles - precompiles dummy bytecode",
  foundationMethods: "dev",
  testCases: ({ context, it, log }) => {
    it({
      id: "T01",
      title: "should return dummy bytecode for every precompiles",
      test: async function () {
        const addresses: Array<`0x${string}`> = [
          "0x0000000000000000000000000000000000000001",
          "0x0000000000000000000000000000000000000002",
          "0x0000000000000000000000000000000000000003",
          "0x0000000000000000000000000000000000000004",
          "0x0000000000000000000000000000000000000005",
          "0x0000000000000000000000000000000000000006",
          "0x0000000000000000000000000000000000000007",
          "0x0000000000000000000000000000000000000008",
          "0x0000000000000000000000000000000000000400",
          "0x0000000000000000000000000000000000000401",
          "0x0000000000000000000000000000000000000402",
          PRECOMPILE_PARACHAIN_STAKING_ADDRESS,
        ];

        const matches = await Promise.all(
          addresses.map(async (address) => {
            const code = await context.viem().getBytecode({ address: address as `0x${string}` });
            return code === DUMMY_REVERT_BYTECODE;
          })
        );

        const failures = matches.filter((match) => !match);
        failures.forEach((failure, index) => {
          log(`Failure at address ${addresses[index]}`);
        });

        expect(
          failures.length,
          "Contract calls to precompile doesn't equal DUMMY_REVERT_BYTECODE"
        ).to.equal(0);
      },
    });

    it({
      id: "T02",
      title: "should revert when dummy bytecode is called",
      test: async function () {
        // we deploy a new contract with the same bytecode to be able to
        // execute the bytecode instead of executing a precompile.
        await context.createBlock(
          createEthersTransaction(context, {
            data: INIT_CODE,
          })
        );

        const contractAddress =
          "0x" +
          keccak256(RLP.encode([ALITH_ADDRESS, 0]), "hex")
            .slice(12)
            .substring(14);
        // check the deployed code by this init code watches what we use for precompiles.
        const code = await context
          .viem()
          .getBytecode({ address: contractAddress as `0x${string}` });
        expect(code).to.equal(DUMMY_REVERT_BYTECODE);

        // try to call contract (with empty data, shouldn't matter)

        const { result } = await context.createBlock(
          createEthersTransaction(context, {
            gas: 12_000_000,
            data: "0x",
            to: contractAddress,
          })
        );
        const receipt = await context
          .viem()
          .getTransactionReceipt({ hash: result!.hash as `0x${string}` });

        expect(receipt.status).to.equal("reverted");
        // 21006 = call cost + 2*PUSH cost
        expect(receipt.gasUsed).to.equal(21006n);
      },
    });
  },
});
