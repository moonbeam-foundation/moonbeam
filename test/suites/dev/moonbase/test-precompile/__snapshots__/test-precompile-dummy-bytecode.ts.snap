// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`🗃️  D012838 Precompiles - precompiles dummy bytecode > 📁  D012838T02 should revert when dummy bytecode is called 1`] = `21006n`;
